3ds2/src/main/java/com/adyen/checkout/adyen3ds2/Adyen3DS2Component.kt

```diff
@@ -0,0 +1,325 @@
+            try {
+                Logger.d(TAG, "initialize 3DS2 SDK")
+                ThreeDS2Service.INSTANCE.initialize(context, configParameters, null, mUiCustomization)
+            } catch (e: SDKRuntimeException) {
+                notifyException(ComponentException("Failed to initialize 3DS2 SDK", e))
+                return@launch
+            } catch (e: SDKAlreadyInitializedException) {
+                // This shouldn't cause any side effect.
+                Logger.w(TAG, "3DS2 Service already initialized.")
+            }
+
+            mTransaction = try {
+                Logger.d(TAG, "create transaction")
+                ThreeDS2Service.INSTANCE.createTransaction(null, fingerprintToken.threeDSMessageVersion)
+            } catch (e: SDKNotInitializedException) {
+                notifyException(ComponentException("Failed to create 3DS2 Transaction", e))
+                return@launch
+            } catch (e: SDKRuntimeException) {
+                notifyException(ComponentException("Failed to create 3DS2 Transaction", e))
+                return@launch
+            }
+
+            val authenticationRequestParameters = mTransaction?.authenticationRequestParameters
+            if (authenticationRequestParameters != null) {
+                val encodedFingerprint = createEncodedFingerprint(authenticationRequestParameters)
+                launch(Dispatchers.Main) {
+                    notifyDetails(createFingerprintDetails(encodedFingerprint))
+                }
+            }
+        }
+    }
+
+    @Throws(ComponentException::class)
+    private fun challengeShopper(activity: Activity, encodedChallengeToken: String) {
+        Logger.d(TAG, "challengeShopper")
+        if (mTransaction == null) {
+            notifyException(Authentication3DS2Exception("Failed to make challenge, missing reference to initial transaction."))
+            return
+        }
+        val decodedChallengeToken = Base64Encoder.decode(encodedChallengeToken)
+        val challengeTokenJson: JSONObject = try {
+            JSONObject(decodedChallengeToken)
+        } catch (e: JSONException) {
+            throw ComponentException("JSON parsing of FingerprintToken failed", e)
+        }
+        val challengeToken = ChallengeToken.SERIALIZER.deserialize(challengeTokenJson)
+        val challengeParameters = createChallengeParameters(challengeToken)
+        try {
+            mTransaction?.doChallenge(activity, challengeParameters, this, DEFAULT_CHALLENGE_TIME_OUT)
+        } catch (e: InvalidInputException) {
+            notifyException(CheckoutException("Error starting challenge", e))
+        }
+    }
+
+    @Throws(ComponentException::class)
+    fun createEncodedFingerprint(authenticationRequestParameters: AuthenticationRequestParameters): String {
+        val fingerprintJson = JSONObject()
+        try {
+            // TODO getMessageVersion is not used?
+            fingerprintJson.put("sdkAppID", authenticationRequestParameters.sdkAppID)
```

Comment:
Minor: you can use `fingerprintJson.apply` or `with(authenticationRequestParameters)` here